### PR TITLE
Updated spec, schema based on goodtables.next

### LINKED ISSUE #3: Reason to use numeric keys and nesting?
@pwalsh 
I'm starting to work on goodtables.next more closely and this spec is a great source for me. But I have a few questions.

First is why to use this numbered keys instead of unique identifier of error. I don't fully understand where this numbers could be used because showing to user things like `error: structure-3` in my opinion less usable then something like `error: structure-duplicate-row`. We have an experience working with linters and I suppose all don't like when they starting to answer to us with some strange numeric codes (ok if you have 200 different error it could be explained). 

Second is `format/structure/schema` nesting. I see here also room for simplification. We have like around 30 errors I suppose we could to write it just is simplest dict form like:

```
unknown-format:
    type: format
    ...
unknown-encoding:
    type: format
    ...
blank-header:
    type: structure
    ...
missing-headers:
    type: schema
    ...
pattern-constraint:
   type: schema
```

It also allow to introduce some **unified language** between users and implementations e.g. there will be `check_missing_headers` or `checkPatternConstraint` (in js). Or in some validator - `validate(checks={'missing-headers': True})`.

Also do you plan tools could use this spec programaticaly? If yes then my comments are even more actual I suppose.

### PR DESCRIPTION
**Pull request by roll**
- fixes #3

---

Here is a question - should be `weight` hard-coded to the spec or it's more application/user level thing (cc @pwalsh @amercader)

### CODE REVIEW

**Comment by pwalsh** [comment]
@roll yes, hard code it. We don't use it in GT anyway right now.

**Comment by pwalsh** [comment]
@roll you've completely dropped name and message. Why?

File: spec.json
Comment (pwalsh) [inline spec.json]: having errors as an array makes just that bit harder to detect duplicate error codes. Seeing as error codes **need** to be unique, why not have an object where the keys are the codes?

File: spec.json
Comment (pwalsh) [inline spec.json]: `general` as a type means little. Why is it better than the current `format`, which at least indicates "it is not the format we expect it is"?

File: spec.json
Comment (pwalsh) [inline spec.json:9]: I am confused by `context`. When I read all the different values for context, I don't really understand what they mean. If anything, they feel like they may be very specific to your implementation of good tables, but perhaps not relevant/useful in a generic data quality spec.

Also, it means there are now two taxonomies: `type` and `context`.

What problem does `context` solve?

File: spec.json
Comment (pwalsh) [inline spec.json]: We are missing two things here that to me seem so critical, I'm not sure why they are not here:
1. `name`: we need a human readable name. GoodTables errors are collated into human-targeted reports. These currently manifest in the GoodTables web UI, and some client libs like in OpenSpending and DataPackagist. Why have you removed name?
2. `message`: like name, we need really clear human readable messages, that can be used by non-technical people to fix errors in their tabular data. Again, the GoodTables UI has many examples of this. In the original spec here, I presented a message string for each error than can easily be manipulated in any programming language that might implement this spec. Why remove it?

File: spec.json
Comment (pwalsh) [inline spec.json]: Also as mentioned in the general comments, we also need to add back weight

**Review by pwalsh — CHANGES_REQUESTED** [review]
This needs discussion and changes before merging.

**Comment by roll** [comment]
@pwalsh 
Thanks.

### name

my bad. will add

### weight

OK. But I have a question - why numbers like 9, 6 etc Not 5, 10 etc Just curious)

### general vs format

I was just trying to find other word then `format` because there is already `format-error` (just one of the type). Also errors like `http-error` don't fit very well with `format` and here could be added other general error we don't know for now. But it's your call. I'm OK with any.

### array vs dict

Dict is better for users but GT requires ordered data structure. Let's handle order on GT level having dict in spec. Anyway using spec literally could be a problem. So it could be better to have some middleware structure on `Inspector` level (to provide order e.g.).

### context

Yes, it's vital for GT but it could be managed on GT level. Other question that in my opinion it's more than just implementation detail. Context (WHERE error is) seems pretty natural to describe tabular data:

dataset

But it's ok to remove from spec.

### message and description

This one is the most questionable for me. Hard-coding message templates will couple spec and implementations too much in my opinion. E.g. I've added many new data to messages for GT.next like `duplicate-row`'s additionally row numbers with the same values etc I see it as an active ongoing process (improving this messages) further. And if we're going to make all changes thru spec it could slow down this improvements. I suppose better to have something like recommended message or just a good description and let implementation handle concrete messages. WDYT?

**Comment by pwalsh** [comment]
@roll is goodtables now using these messages? Apart from that, all looks good.

**Comment by roll** [comment]
@pwalsh 
I'll next make a separate PR to update GT to work with this spec (and will use this messages as templates). So it's the last thing.

**Review by pwalsh — APPROVED** [review]
can merge

**Pull request merged**
